### PR TITLE
Add documentation about in-place methods

### LINKED ISSUE #336: Add documentation about in-place solvers

### PR DESCRIPTION
**Pull request by amontoison**
fix #336

### CODE REVIEW

File: docs/src/matrix-free.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-## Matrix-free operators
+## [Matrix-free operators](@id matrix-free)
```
Comment (dpo) [inline docs/src/matrix-free.md:1]: I prefer "factorization-free" to "matrix-free" to describe a *method*. "Factorization-free operator" and "matrix-free operator" don't make much sense though. How about "Abstract operators"?

Below, I would change "All methods are matrix free" to "All methods are factorization free".

File: docs/src/inplace.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
+## In-place methods
+
+All solvers in Krylov.jl have in-place method which ends with `!`.
```
Comment (dpo) [inline docs/src/inplace.md:3]:
```suggestion
All solvers in Krylov.jl have an in-place variant implemented in a method whose name ends with `!`.
```

File: docs/src/inplace.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+A workspace (`KrylovSolver`) that contains the storage needed by a Krylov method can be used to solve multiple linear systems that have the same size and floating-point precision.
```
Comment (dpo) [inline docs/src/inplace.md:4]:
```suggestion
A workspace (`KrylovSolver`) that contains the storage needed by a Krylov method can be used to solve multiple linear systems that have the same dimensions in the same floating-point precision.
```

File: docs/src/inplace.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+Each `KrylovSolver` has two constructors:
+
+```constructors
+XyzSolver(A, b)
+XyzSolver(n, m, S)
+```
+
+`Xyz` is the name of the Krylov method with lowercase letters except its first one (`Cg`, `Minres`, `Lsmr`, `Bicgstab`, ...).
+Given an operator `A` and a right-hand side `b`, you can create a `KrylovSolver` based on the size of `A` and the type of `b` or explicitly give the dimensions `(n, m)` and the storage type `S`.
```
Comment (dpo) [inline docs/src/inplace.md:13]:
```suggestion
Given an operator `A` and a right-hand side `b`, you can create a `KrylovSolver` based on the size of `A` and the type of `b` or explicitly give the dimensions `(m, n)` and the storage type `S`.
```

File: docs/src/inplace.md [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+For example, `S = Vector{Float64}` if you want to solve linear systems in double precision with CPUs and `S = CuVector{Float32}` if you want to solve linear systems in single precision with Nvidia GPUs.
```
Comment (dpo) [inline docs/src/inplace.md:15]:
```suggestion
For example, use `S = Vector{Float64}` if you want to solve linear systems in double precision on the CPU and `S = CuVector{Float32}` if you want to solve linear systems in single precision on an Nvidia GPU.
```

File: docs/src/inplace.md [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+!!! note
+    `DiomSolver`, `DqgmresSolver` and `CgLanczosShiftSolver` require an additional argument.
```
Comment (dpo) [inline docs/src/inplace.md:18]: Which is what?

File: docs/src/inplace.md [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+The workspace is always the first argument of the in-place methods:
+
+```solvers
+minres_solver = MinresSolver(n, n, Vector{Float64})
+x, stats = minres!(minres_solver, A1, b1)
+
+dqgmres_solver = DqgmresSolver(n, n, memory, Vector{BigFloat})
+x, stats = dqgmres!(dqgmres_solver, A2, b2)
+
+lsqr_solver = LsqrSolver(n, m, CuVector{Float32})
```
Comment (dpo) [inline docs/src/inplace.md:29]: `n, m` should be `m, n`. Nobody gives the number of columns first.

File: docs/src/inplace.md [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+x, stats = lsqr!(lsqr_solver, A3, b3)
+```
+
```
Comment (dpo) [inline docs/src/inplace.md:32]: I think you could describe here what kind of operations an "operator" should support.

File: docs/src/inplace.md [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+dqgmres_solver = DqgmresSolver(n, n, memory, Vector{BigFloat})
+x, stats = dqgmres!(dqgmres_solver, A2, b2)
+
+lsqr_solver = LsqrSolver(n, m, CuVector{Float32})
+x, stats = lsqr!(lsqr_solver, A3, b3)
+```
+  
+## Examples
+
+We illustrate the use of in-place Krylov solvers for two well-known optimization methods.
```
Comment (dpo) [inline docs/src/inplace.md:35]:
```suggestion
We illustrate the use of in-place Krylov solvers with two well-known optimization methods.
```

File: docs/src/inplace.md [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+The details of the optimization methods are described in the section about [Matrix-free operators](@ref matrix-free).
```
Comment (dpo) [inline docs/src/inplace.md:36]: See comment below.

File: docs/src/inplace.md [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+Given an operator `A` and a right-hand side `b`, you can create a `KrylovSolver` based on the size of `A` and the type of `b` or explicitly give the dimensions `(n, m)` and the storage type `S`.
+
+For example, `S = Vector{Float64}` if you want to solve linear systems in double precision with CPUs and `S = CuVector{Float32}` if you want to solve linear systems in single precision with Nvidia GPUs.
+
+!!! note
+    `DiomSolver`, `DqgmresSolver` and `CgLanczosShiftSolver` require an additional argument.
+
+The workspace is always the first argument of the in-place methods:
+
+```solvers
```
Comment (dpo) [inline docs/src/inplace.md:22]: Is this missing an `@`?

File: docs/src/inplace.md [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+
+We illustrate the use of in-place Krylov solvers for two well-known optimization methods.
+The details of the optimization methods are described in the section about [Matrix-free operators](@ref matrix-free).
+
+### Example 1: Newton's Method for convex optimization
+
+```@newton
+using Krylov
+
+function newton(f, ∇f, ∇²f, x₀; itmax = 200, tol = 1e-8)
```
Comment (dpo) [inline docs/src/inplace.md:43]:
```suggestion
function newton(∇f, ∇²f, x₀; itmax = 200, tol = 1e-8)
```

File: docs/src/inplace.md [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+lsqr_solver = LsqrSolver(n, m, CuVector{Float32})
+x, stats = lsqr!(lsqr_solver, A3, b3)
+```
+  
+## Examples
+
+We illustrate the use of in-place Krylov solvers for two well-known optimization methods.
+The details of the optimization methods are described in the section about [Matrix-free operators](@ref matrix-free).
+
+### Example 1: Newton's Method for convex optimization
```
Comment (dpo) [inline docs/src/inplace.md:38]:
```suggestion
### Example 1: Newton's method for convex optimization without linesearch
```

File: docs/src/inplace.md [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+        
+        iter += 1
+        solved = norm(gx) ≤ tol
+        tired = iter ≥ itmax
+    end
+    return x
+end
+```
+
+### Example 2: The Gauss-Newton Method for Nonlinear Least Squares
```
Comment (dpo) [inline docs/src/inplace.md:71]:
```suggestion
### Example 2: The Gauss-Newton method for nonlinear least squares without linesearch
```

**Review by dpo — CHANGES_REQUESTED** [review]

File: docs/src/inplace.md (hunk 8 above)
Comment (amontoison) [reply docs/src/inplace.md:32]: We did it last year :)
https://juliasmoothoptimizers.github.io/Krylov.jl/stable/matrix-free/

**Comment by dpo** [comment]
BTW the logo looks weird in the docs

File: docs/src/inplace.md (hunk 8 above)
Comment (dpo) [reply docs/src/inplace.md:32]: Ok. Didn't read the docs ;-)

**Comment by amontoison** [comment]
> BTW the logo looks weird in the docs

Already fixed : https://github.com/JuliaSmoothOptimizers/Krylov.jl/pull/342

**Comment by amontoison** [comment]
Thanks for the comments Dominique !
I updated the PR to take them into account.

**Pull request merged**